### PR TITLE
Add Keras/Class Activation Map article

### PR DESCRIPTION
**Pull request by ReeceStevens**
This isn't yet complete, but I'm putting up a PR so we can leave inline comments.

### CODE REVIEW

File: 10x/_posts/2017-12-07-pretrained-models-with-keras.md [hunk 1]
```diff
@@ -0,0 +1,588 @@
...
+import keras
+from keras.layers.core import Dense, Flatten, Dropout
+from keras.layers.normalization import BatchNormalization
+from keras.models import Model
+
+IMG_SIZE = (256, 256)
+
+def get_model(pretrained_model, all_character_names):
+    if pretrained_model == 'inception':
+        pretrained_model = keras.applications.inception_v3.InceptionV3(include_top=False, input_shape=(*IMG_SIZE, 3), weights='imagenet')
```
Comment (johndgiese) [inline 10x/_posts/2017-12-07-pretrained-models-with-keras.md]: We should try to get all our code within the 80 (or worst case 100) character maximum width.  E.g. perhaps we can use `from keras.applications.inceptions_v3 import InceptionV3`?  I think this is especially important for any code that will be in the main article.

File: 10x/_posts/2017-12-07-pretrained-models-with-keras.md [hunk 2]
```diff
@@ -0,0 +1,588 @@
...
+the picture is of a dog, cat, or bird? Classification algorithms have usually
+relied upon carefully hand-crafted image features to identify images in such a
+manner. Thanks to recent developments in machine learning, however, we now have
+another option: we can train a deep learning model.
+
+Deep learning models such as convolutional neural networks provide a new
+way to solve previously intractable problems. By training a network
+of neurons on a large quantity of labelled image data, we can build a
+probabalistic model that will classify new images into categories with a high
+degree of accuracy. The only drawback, however, is that for sufficiently
```
Comment (johndgiese) [inline 10x/_posts/2017-12-07-pretrained-models-with-keras.md]: > The only drawback, however, ...

Perhaps we should say "A major drawback ..." since "only" is pretty strong.

File: 10x/_posts/2017-12-07-pretrained-models-with-keras.md [hunk 3]
```diff
@@ -0,0 +1,588 @@
...
+author: "Reece Stevens"
+layout: "10x"
+---
+
+At Innolitics, we work in a wide variety of medical imaging contexts. Often in
+our work with clients, we find that a decision has to be made based on
+information encoded in an image or set of images. Depending on the type of
+information we need, extracting meaning can be quite a difficult task to
+automate. One instance where this happens frequently is in image
+classification: if a computer is given a picture of an animal, can it tell if
```
Comment (johndgiese) [inline 10x/_posts/2017-12-07-pretrained-models-with-keras.md]: It would be nice if we could replace the animal example, with a medical example... e.g. our CIRS classification stage.

File: 10x/_posts/2017-12-07-pretrained-models-with-keras.md [hunk 4]
```diff
@@ -0,0 +1,588 @@
...
+image-recognition competition that attempts to sort images into 1000 different
+categories of everyday objects. Their model, [Inception v3][inception], is
+highly complex-- according to the GitHub instructions for training the model
+from scratch, [it can take days to weeks even with a multi-GPU hardware
+setup][inception-from-scratch]. If you don't have access to a high-quality GPU,
+that training time can jump even higher. The hardware and time constraints of
+model training often serve as a signficant barrier to entry in the machine
+learning field.
+
+That's where pre-trained models come in. Rather than attempt to train an entire
```
Comment (johndgiese) [inline 10x/_posts/2017-12-07-pretrained-models-with-keras.md]: @ReeceStevens I think splitting this into two articles would be a good idea:

Article 1: Pretraining with Keras
Article 2: CAM visualization stuff

File: 10x/_posts/2017-12-07-pretrained-models-with-keras.md [hunk 5]
```diff
@@ -0,0 +1,588 @@
...
+                                                          save_best_only=False,
+                                                          save_weights_only=False,
+                                                          mode='auto',
+                                                          period=1)
+
+    data_generator = DataGenerator(args.data_dir)
+    model = get_model(args.pretrained_model, data_generator.encoder.all_character_names)
+
+    model.fit_generator(
+        data_generator.batch_generator('train', batch_size=BATCH_SIZE),
```
Comment (johndgiese) [inline 10x/_posts/2017-12-07-pretrained-models-with-keras.md]: I like this indent style better (esp of articles because people may be reading on their phone) than the one used:

https://github.com/innolitics/innolitics.github.io/pull/18/files#diff-d0e3f24a9e6c7bb1eda83195bf72a8feR349

or 

https://github.com/innolitics/innolitics.github.io/pull/18/files#diff-d0e3f24a9e6c7bb1eda83195bf72a8feR353

In any event, we should definitely be consistent.

File: 10x/_posts/2017-12-07-pretrained-models-with-keras.md (hunk 2 above)
Comment (ReeceStevens) [reply 10x/_posts/2017-12-07-pretrained-models-with-keras.md]: That's a really good point-- I was trying to distill the introduction down into the problem we were solving with pre-trained models, but that's definitely not the only issue with deep learning and it's a very big assumption to say we have all the data necessary. I'll definitely update that :)

File: 10x/_posts/2017-12-07-pretrained-models-with-keras.md (hunk 3 above)
Comment (ReeceStevens) [reply 10x/_posts/2017-12-07-pretrained-models-with-keras.md]: Perhaps we can re-work this to start with a basic animal example and then link that to an analogous medical example? The classification stage of CIRS is a great representative problem but it might be hard to visualize right off the bat (I know I struggled to visualize it until I saw pictures).

Alternatively, are there any images from CIRS that we would be able to use as an example? It might be cool to have a visual instance of the problem. Not mission critical, but it could be useful.

File: 10x/_posts/2017-12-07-pretrained-models-with-keras.md (hunk 5 above)
Comment (ReeceStevens) [reply 10x/_posts/2017-12-07-pretrained-models-with-keras.md]: Sure thing, I'll update that.

File: 10x/_posts/2017-12-07-pretrained-models-with-keras.md (hunk 3 above)
Comment (yujanshrestha) [reply 10x/_posts/2017-12-07-pretrained-models-with-keras.md]: Maybe a more relatable medical example:
"If a computer is given a picture of a skin lesion, can it tell if it is non-cancerous, basal cell carcinoma, squamous cell carcinoma, or a melanoma?"

File: 10x/_posts/2017-12-07-pretrained-models-with-keras.md [hunk 6]
```diff
@@ -0,0 +1,588 @@
...
+
+At Innolitics, we work in a wide variety of medical imaging contexts. Often in
+our work with clients, we find that a decision has to be made based on
+information encoded in an image or set of images. Depending on the type of
+information we need, extracting meaning can be quite a difficult task to
+automate. One instance where this happens frequently is in image
+classification: if a computer is given a picture of an animal, can it tell if
+the picture is of a dog, cat, or bird? Classification algorithms have usually
+relied upon carefully hand-crafted image features to identify images in such a
+manner. Thanks to recent developments in machine learning, however, we now have
```
Comment (yujanshrestha) [inline 10x/_posts/2017-12-07-pretrained-models-with-keras.md]: Perhaps something more concrete: "Thanks to recent advancements in computational hardware and the explosion of available digitized data"

File: 10x/_posts/2017-12-07-pretrained-models-with-keras.md [hunk 7]
```diff
@@ -0,0 +1,588 @@
...
+another option: we can train a deep learning model.
+
+Deep learning models such as convolutional neural networks provide a new
+way to solve previously intractable problems. By training a network
+of neurons on a large quantity of labelled image data, we can build a
+probabalistic model that will classify new images into categories with a high
+degree of accuracy. The only drawback, however, is that for sufficiently
+complex problems it is difficult to have enough computing power to train the
```
Comment (yujanshrestha) [inline 10x/_posts/2017-12-07-pretrained-models-with-keras.md]: "A significant drawback, however, is that for sufficiently complicated problems it is difficult to have enough computing power to train the model in any reasonable amount of time"

File: 10x/_posts/2017-12-07-pretrained-models-with-keras.md [hunk 8]
```diff
@@ -0,0 +1,588 @@
...
+model in any reasonable amount of time.
+
+For example: the Google Brain team built an image classifier model to compete
+in the [ImageNet Large Scale Visual Recognition Challenge][imagenet], an
+image-recognition competition that attempts to sort images into 1000 different
+categories of everyday objects. Their model, [Inception v3][inception], is
+highly complex-- according to the GitHub instructions for training the model
+from scratch, [it can take days to weeks even with a multi-GPU hardware
+setup][inception-from-scratch]. If you don't have access to a high-quality GPU,
```
Comment (yujanshrestha) [inline 10x/_posts/2017-12-07-pretrained-models-with-keras.md]: "If you don't have access to a cluster of GPUs,"

File: 10x/_posts/2017-12-07-pretrained-models-with-keras.md [hunk 9]
```diff
@@ -0,0 +1,588 @@
...
+that training time can jump even higher. The hardware and time constraints of
+model training often serve as a signficant barrier to entry in the machine
+learning field.
```
Comment (yujanshrestha) [inline 10x/_posts/2017-12-07-pretrained-models-with-keras.md]: 👍

File: 10x/_posts/2017-12-07-pretrained-models-with-keras.md [hunk 10]
```diff
@@ -0,0 +1,588 @@
...
+
+That's where pre-trained models come in. Rather than attempt to train an entire
+image recognition neural network, we instead stand on the shoulders of giants
+and use networks trained by other research groups. It is possible for us to use
+portions of high-quality ImageNet models to do the heavy lifting and help us
+create our own image classifier-- and this task is made even easier with the
+help of [Keras][keras], a deep-learning library for python that makes creating
```
Comment (yujanshrestha) [inline 10x/_posts/2017-12-07-pretrained-models-with-keras.md]: I think we should capitalize Python

File: 10x/_posts/2017-12-07-pretrained-models-with-keras.md [hunk 11]
```diff
@@ -0,0 +1,588 @@
...
+    resized_image_pixels = scipy.misc.imresize(image_pixels, IMG_SIZE)
+    image_basepath, _ = os.path.splitext(image_path)
+    np.savez(image_basepath+'.npz', pixels=resized_image_pixels)
+
+```
+
+After we normalize the image dimensions, our next task is to partition the
+dataset into training, validation, and testing sets. _Training_ data is used
+during the training of the model; _validation_ data is used to gauge the accuracy
+of the model during training; and _testing_ data is used at the end of
```
Comment (yujanshrestha) [inline 10x/_posts/2017-12-07-pretrained-models-with-keras.md]: `_validation_ data is used to detect overfitting during training`

File: 10x/_posts/2017-12-07-pretrained-models-with-keras.md (hunk 6 above)
Comment (ReeceStevens) [reply 10x/_posts/2017-12-07-pretrained-models-with-keras.md]: Ohh, yeah that's much better. +1

File: 10x/_posts/2017-12-07-pretrained-models-with-keras.md (hunk 11 above)
Comment (ReeceStevens) [reply 10x/_posts/2017-12-07-pretrained-models-with-keras.md]: Perhaps `_validation_ data is used to determine model accuracy on a separate dataset during training`? I think it might be confusing to claim it "detects overfitting" since we later propose the idea that InceptionV3 scores more highly in the validation data's accuracy metric due to possible overfitting.
